### ISSUE TITLE
type-declaration-immutability: ignorePattern doesn't work on member name

### ISSUE THREAD

**Issue by adrian-gierakowski** [eslint-functional/eslint-plugin-functional#599]
## Bug Report

`ignorePattern` option of type-declaration-immutability doesn't work on member names, contrary to what is advertised  [here](https://github.com/eslint-functional/eslint-plugin-functional/blob/f441a12a880b07c36d8076065ea6796d0b081f6a/docs/rules/options/ignore-pattern.md)

### Expected behavior

given following config:

```js
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  plugins: ['@typescript-eslint', 'functional'],
    rules: {
      'functional/type-declaration-immutability': [
        'error',
        {
          ignorePattern: ['^mutable'],
        }
      ],
    },
    parserOptions: {
      project: ['./tsconfig.lint.json'],
    },
}
```

there should be no linter error for the following code:

```ts
type Person = {
  readonly name: string;
  mutableAge: number;
};
```

### Actual behavior

The following error is raised:

```
1:6  error  This type is declare to have an immutability of at least "Immutable" (actual: "Mutable")  functional/type-declaration-immutability
```

Note that the following does not cause an error, which means that the `ignorePattern` is in force, but only matches the type name, not it's members.

```ts
type mutablePerson = {
  readonly name: string;
  age: number;
};
```

### Steps to reproduce

Run eslint on the above code using the above config.

### Proposed changes

Mutability checks for members which names match `ignorePattern` should be skipped

**Comment by RebeccaStevens** [comment]
The `ignorePattern` only works on the alias name, not on any of its members (#467).

If we were to add support for ignoring members, we'd likely need do it in a new option.
Though, I don't think it's very likely that this will be add.
In the example given, I would not be in favor of letting `Person` be considered immutable when it is very much mutable.

**Comment by adrian-gierakowski** [comment]
Right, so maybe [this page](https://github.com/eslint-functional/eslint-plugin-functional/blob/f441a12a880b07c36d8076065ea6796d0b081f6a/docs/rules/options/ignore-pattern.md) should be deleted, or edited?

**Comment by RebeccaStevens** [comment]
> The example above was taken verbatim from your own documentation, so maybe [this page](https://github.com/eslint-functional/eslint-plugin-functional/blob/f441a12a880b07c36d8076065ea6796d0b081f6a/docs/rules/options/ignore-pattern.md) should be deleted, or edited?

Yeah, I should have re-written that file with the release of v5.

The reasoning behind my last comment and closing this issue was based off the fact that with v5, the rules `type-declaration-immutability` and `prefer-immutable-types` take a different philosophy to the v1-v4 `prefer-readonly-type`.
I don't think I've fully documented this philosophy properly and how it differs from the previous one. I'm not exactly sure how to best document it though.

**Comment by RebeccaStevens** [comment]
Note: Once [@typescript-eslint/type-utils](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/type-utils) v6 comes out, I'll be making some improvements to [is-immutable-type](https://github.com/RebeccaStevens/is-immutable-type) using the new functionality coming (#575 and #577 are waiting for this).
This update will likely create some breaking changes, so I'll probably a make new major release of this plugin then (#598 will likely get included with in).

**Comment by rickmed** [comment]
So let's say I have a class (or object). Is there any way to allow this?

```ts
class Hey {
    _mut_array = []
}

const hey = new Hey()

hey._mut_array.push(1)   // "Modifying an array is not allowed."
```

**Comment by RebeccaStevens** [comment]
@rickmed I think you replied to the wrong thread. https://github.com/eslint-functional/eslint-plugin-functional/issues/691#issuecomment-1646747915

**Comment by rickmed** [comment]
@RebeccaStevens don't think so. The other issue is about const/let functionality.

**Comment by RebeccaStevens** [comment]
Ahh, sorry. I thought this was just a follow up question for over there. I'll move my reply here.

**Comment by RebeccaStevens** [comment]
> So let's say I have a class (or object). Is there any way to do this?
> 
> ```ts
> class Hey {
>     mut_array = []
> }
> 
> const hey = new Hey()
> 
> hey.mut_array.push(1)   // currently, it complains with "Modifying an array is not allowed."
> ```

You can use the [`ignoreAccessorPattern` option](https://github.com/eslint-functional/eslint-plugin-functional/blob/main/docs/rules/immutable-data.md#ignoreaccessorpattern) for this.

This should do what you want:

```jsonc
{
  "ignoreAccessorPattern": "**.mut_*.**"
}
```
